src/core/ddsc/src/dds_reader.c

```diff
@@ -602,16 +602,17 @@ static dds_entity_t dds_create_reader_int (dds_entity_t participant_or_subscribe
   dds_entity_init_complete (&rd->m_entity);
 
 #ifdef DDS_HAS_SHM
-  rqos->shared_memory.enabled = dds_reader_support_shm(&gv->config, rqos);
-  rqos->present |= QP_SHARED_MEMORY;
+  if (!dds_reader_support_shm(&gv->config, rqos))
+    rqos->ignore_locator_type |= NN_LOCATOR_KIND_SHEM;
```

Comment:
I like the changing the "shared memory" flag to a set of supported locator kinds. There are, however, two (at least 🙂) problems with this:
* The first is minor and a matter of taste: I think I'd find it more elegant if it was a set of enabled locator kinds instead. The downside is of course that the default can no longer be simply 0. It could be that in the end that works out as messier than an ignored set.
* The second is rather major: locator kinds are 32-bit integers with no range limitation, and so a bitset won't do. What I have considered at times is to translate between internal and external locator kinds, because the code only deals with a handful and it generally is attractive to be able to use small integers. Being able to represent them in a small bitmask being one, efficiently looking up the `dds_tran_factory` supporting it being another.

As far as that translation goes, I think it'd be pretty straightforward by now. It already has conversion functions in place (in/near https://github.com/eclipse-cyclonedds/cyclonedds/blob/3108aff3ee8c8b4cdb418441d1fe0925de3ef273/src/core/ddsi/src/ddsi_plist.c#L376), and no locators ever go in/out without going through these functions. So implementing this translation should be no more than a limited bit of refactoring.

Reply:
I was thinking of using "allowed" locator kinds first, but since this caused issues with initializing the default, I used the inverse, so a "disallowed", mask first.

I will look into the conversion functions, since this will alow this mechanism to become more generic.